providers/gitlab.go

```diff
@@ -15,11 +18,49 @@ import (
 type GitLabProvider struct {
 	*ProviderData
 
-	Groups               []string
-	Verifier             *oidc.IDTokenVerifier
+	Groups       []string
+	Projects     []*GitlabProject
+	EmailDomains []string
+
 	AllowUnverifiedEmail bool
 }
 
+// GitlabProject represents a Gitlab project constraint entity
+type GitlabProject struct {
+	Name        string
+	AccessLevel int
+}
+
+// Creates a new GitlabProject struct from project string formatted as namespace/project=accesslevel
+// if no accesslevel provided, use the default one
+func gitlabProjectFromString(project string) (*GitlabProject, error) {
```

Comment:
NIT: this method name feels overly verbose (I can see its a string based on the arguments signature).

This is essentially an initializer right? Could be `newGitlabProject` to correspond to the struct declared immediately above it.